spec/java/util/StringBuilder.main.lsl

```diff
@@ -0,0 +1,1022 @@
+    {
+        _checkOffset(dstOffset);
+        var s: String = action OBJECT_TO_STRING(ii);
+
+        if (s == null)
+            s = "null";
+
+        val len: int = action CALL_METHOD(s, "length", []);
+        _insertCharSequence(dstOffset, s, len, 0, len);
+
+        result = self;
+    }
+
+
+    fun *.insert (@target self: StringBuilder, dstOffset: int, l: long): StringBuilder
+    {
+        _checkOffset(dstOffset);
+        var s: String = action OBJECT_TO_STRING(l);
+
+        if (s == null)
+            s = "null";
+
+        val len: int = action CALL_METHOD(s, "length", []);
+        _insertCharSequence(dstOffset, s, len, 0, len);
+
+        result = self;
+    }
+
+
+    fun *.lastIndexOf (@target self: StringBuilder, str: String): int
+    {
+        result = action DEBUG_DO("this.storage.lastIndexOf(str)");
+    }
+
+
+    fun *.lastIndexOf (@target self: StringBuilder, str: String, fromIndex: int): int
+    {
+        result = action DEBUG_DO("this.storage.lastIndexOf(str, fromIndex)");
+    }
+
+
+    fun *.replace (@target self: StringBuilder, start: int, end: int, str: String): StringBuilder
+    {
+        if (end > this.length)
+            end = this.length;
+
+        _checkRangeSIOOBE(start, end, this.length);
+
+        val strLength: int = action CALL_METHOD(str, "length", []);
+
+        val newLength: int = this.length + strLength - (end - start);
+        var i: int = 0;
+        var strIndex: int = 0;
+        val newStr: array<char> = action ARRAY_NEW("char", newLength);
+
+        action LOOP_FOR(i, 0, newLength, +1, _replace_loop(i, strIndex, start, end, newStr, str, strLength));
+
+        this.storage = action OBJECT_TO_STRING(newStr);
+        this.length = newLength;
+        result = self;
+    }
+
+
+    @Phantom proc _replace_loop (i: int, strIndex: int, start: int, end: int, newStr: array<char>, str: String, strLength: int): void
+    {
+        if (i < start)
+        {
+            newStr[i] = action CALL_METHOD(this.storage, "charAt", [i]);
+        }
+        else if (i < start + strLength)
+        {
+            newStr[i] = action CALL_METHOD(str, "charAt", [strIndex]);
+            strIndex += 1;
+        }
+        else
+        {
+            newStr[i] = action CALL_METHOD(this.storage, "charAt", [end]);
+            end += 1;
+        }
+    }
+
+
+    fun *.reverse (@target self: StringBuilder): StringBuilder
+    {
+        var n: int = this.length - 1;
+        var i: int = (n-1) / 2;
+        var k: int = -1;
+        val newStr: array<char> = action ARRAY_NEW("char", this.length);
+        action LOOP_FOR(i, 0, -1, -1, _reverse_loop(i, k, n, newStr));
+
+        this.storage = action OBJECT_TO_STRING(newStr);
+        result = self;
+    }
+
+
+    @Phantom proc _reverse_loop (i: int, k: int, n: int, newStr: array<char>): void
+    {
+        k = n - i;
+        newStr[k] = action CALL_METHOD(this.storage, "charAt", [i]);
```

Comment:
I don't think swaps are better than just creating a new array and copying all characters into that (also using just plus and minus one instead of *complex* math like full fledged integer subtraction).

Reply:
I suggest the following instead:
```dart
fun *.reverse (@target self: StringBuilder): StringBuilder
{
    if (this.length != 0)
    {
        action ASSUME(this.length > 0);
        // serialize current state of storage string
        val oldStorage: array<char> = action CALL_METHOD(this.storage, "toCharArray", []);

        // prepare a new serialized but processed version of the storage
        val newStorage: array<char> = action ARRAY_NEW("char", this.length);
        action ASSUME(action ARRAY_SIZE(newStorage) == this.length);
        action ASSUME(action ARRAY_SIZE(oldStorage) == action ARRAY_SIZE(newStorage));

        var j: int = this.length - 1;
        var i: int = 0;
        action LOOP_FOR(
            i, 0, this.length, +1,
            _reverse_loop(i, oldStorage, j, newStorage)
        );

        // replace the old state with a reversed (buffer) version
        this.storage = action OBJECT_TO_STRING(newStorage);
    }
    result = self;
}

@Phantom proc _reverse_loop (i: int, oldStorage: array<char>, j: int, newStorage: array<char>): void
{
    newStorage[j] = oldStorage[i];
    j -= 1;
}
```